File: src/python/nest_geometry.py
"""Bird's Nest hyperbolic paraboloid geometry engine.

Computes pillar positions and roof segments using the doubly-ruled
surface property of the hyperbolic paraboloid.

Parameters (prototype-verified):
    HP Surface:   z = C*(x^2/AHP^2 - y^2/BHP^2) + Z_OFFSET
    Outer ellipse boundary (ground projection): (x/A_OUT)^2 + (y/B_OUT)^2 = 1
    Inner ellipse boundary (roof opening):     (x/A_IN)^2  + (y/B_IN)^2  = 1
"""

import numpy as np

# ---- Global constants (prototype-verified) ----
AHP = 80.0
BHP = 72.0
C = 18.0
Z_OFFSET = 23.2
A_OUT = 60.0
B_OUT = 53.5
A_IN = 34.4
B_IN = 22.4
NU = 50
NV = 50
U_RANGE = (-1.06, 1.06)
V_RANGE = (-1.06, 1.06)


def hyperboloid_z(x, y):
    """HP surface height at (x, y)."""
    return C * (x**2 / AHP**2 - y**2 / BHP**2) + Z_OFFSET


def line_ellipse_intersect(m, d, a_ell, b_ell):
    """Intersection of line y = m*x + d with ellipse (x/a)^2 + (y/b)^2 = 1.

    Returns list of (x, y) tuples: 0, 1, or 2 points.
    """
    A = 1.0 / a_ell**2 + m**2 / b_ell**2
    B = 2.0 * m * d / b_ell**2
    Cc = d**2 / b_ell**2 - 1.0
    disc = B**2 - 4.0 * A * Cc

    if disc < 0.0:
        return []
    elif disc < 1e-12:
        x = -B / (2.0 * A)
        return [(x, m * x + d)]
    else:
        sqrt_disc = np.sqrt(disc)
        x1 = (-B + sqrt_disc) / (2.0 * A)
        x2 = (-B - sqrt_disc) / (2.0 * A)
        return [(x1, m * x1 + d), (x2, m * x2 + d)]


def point_in_ellipse(x, y, a_ell, b_ell):
    """True if (x, y) is strictly inside the ellipse."""
    return (x**2 / a_ell**2 + y**2 / b_ell**2) < 1.0


def _process_ruling(m, d, pillars, roof_segments):
    """Process one ruling line: intersect ellipses, connect valid segments."""
    pts_out = line_ellipse_intersect(m, d, A_OUT, B_OUT)
    pts_in = line_ellipse_intersect(m, d, A_IN, B_IN)

    all_pts = []  # list of (x, y, type, t_param)
    for (x, y) in pts_out:
        all_pts.append((x, y, 1, x))  # type=1: outer
    for (x, y) in pts_in:
        all_pts.append((x, y, 2, x))  # type=2: inner

    if len(all_pts) < 2:
        return

    all_pts.sort(key=lambda pt: pt[3])  # sort by t_param

    for j in range(len(all_pts) - 1):
        x1, y1, t1, _ = all_pts[j]
        x2, y2, t2, _ = all_pts[j + 1]

        if t1 == 2 and t2 == 2:
            continue  # skip inner-inner (inside opening)

        mid_x = (x1 + x2) / 2.0
        mid_y = (y1 + y2) / 2.0
        if point_in_ellipse(mid_x, mid_y, A_IN, B_IN):
            continue  # segment crosses the opening

        z1 = hyperboloid_z(x1, y1)
        z2 = hyperboloid_z(x2, y2)
        pillars.append((x1, y1, 0.0, z1))
        pillars.append((x2, y2, 0.0, z2))
        roof_segments.append(((x1, y1, z1), (x2, y2, z2)))


def compute_geometry():
    """Compute all nest geometry.

    Returns dict with:
        pillars:       np.ndarray [Nx4] (x, y, z_base, z_top)
        roof_segments: list of ((x1,y1,z1),(x2,y2,z2))
        ellipse_outer: np.ndarray [2x200]
        ellipse_inner: np.ndarray [2x200]
        u_count, v_count: int
    """
    pillars = []        # list of (x, y, z_base, z_top)
    roof_segments = []  # list of ((x1,y1,z1),(x2,y2,z2))

    slope_u = BHP / AHP    # +0.9
    slope_v = -BHP / AHP   # -0.9

    # ---- u-family: x/AHP - y/BHP = u ----
    u_vals = np.linspace(U_RANGE[0], U_RANGE[1], NU)
    for u in u_vals:
        d = -slope_u * AHP * u
        _process_ruling(slope_u, d, pillars, roof_segments)

    # ---- v-family: x/AHP + y/BHP = v ----
    v_vals = np.linspace(V_RANGE[0], V_RANGE[1], NV)
    for v in v_vals:
        d = -slope_v * AHP * v
        _process_ruling(slope_v, d, pillars, roof_segments)

    # ---- ellipse outlines ----
    theta = np.linspace(0.0, 2.0 * np.pi, 200)
    ellipse_outer = np.array([A_OUT * np.cos(theta), B_OUT * np.sin(theta)])
    ellipse_inner = np.array([A_IN * np.cos(theta), B_IN * np.sin(theta)])

    return {
        'pillars': np.array(pillars),
        'roof_segments': roof_segments,
        'ellipse_outer': ellipse_outer,
        'ellipse_inner': ellipse_inner,
        'u_count': NU,
        'v_count': NV,
    }

File: src/matlab/nest_geometry.m
function G = nest_geometry(p)
% NEST_GEOMETRY 计算鸟巢双曲抛物面动画的全部几何数据 (Iteration 2)
if nargin < 1 || isempty(p) || ~isstruct(p)
    ahp = 80; bhp = 72; c_val = 18; z_offset = 23.2;
    a_out = 60; b_out = 53.5; a_in = 34.4; b_in = 22.4;
    Nu = 50; Nv = 50;
    u_range = [-1.06, 1.06]; v_range = [-1.06, 1.06];
else
    ahp = p.ahp; bhp = p.bhp; c_val = p.c; z_offset = p.z_offset;
    a_out = p.a_out; b_out = p.b_out; a_in = p.a_in; b_in = p.b_in;
    Nu = p.Nu; Nv = p.Nv;
    u_range = p.u_range; v_range = p.v_range;
end
hpZ = @(x,y) c_val*(x.^2/ahp^2 - y.^2/bhp^2) + z_offset;
pillars = []; roof_segments = {};
slope_u = bhp/ahp; slope_v = -bhp/ahp;
u_vals = linspace(u_range(1), u_range(2), Nu);
for ui = 1:Nu
    d = -slope_u*ahp*u_vals(ui);
    [pillars, roof_segments] = processRuling(slope_u, d, a_out, b_out, a_in, b_in, hpZ, pillars, roof_segments);
end
v_vals = linspace(v_range(1), v_range(2), Nv);
for vi = 1:Nv
    d = -slope_v*ahp*v_vals(vi);
    [pillars, roof_segments] = processRuling(slope_v, d, a_out, b_out, a_in, b_in, hpZ, pillars, roof_segments);
end
theta = linspace(0, 2*pi, 200);
G.ellipse_outer = [a_out*cos(theta); b_out*sin(theta)];
G.ellipse_inner = [a_in*cos(theta); b_in*sin(theta)];
G.pillars = pillars; G.roof_segments = roof_segments;
G.u_count = Nu; G.v_count = Nv;
end

function [pillars, roof_segments] = processRuling(m, d, a_out, b_out, a_in, b_in, hpZ, pillars, roof_segments)
pts_out = lineEllipseIntersect(m, d, a_out, b_out);
pts_in  = lineEllipseIntersect(m, d, a_in, b_in);
all_pts = [];
if ~isempty(pts_out)
    for j = 1:size(pts_out,1), all_pts = [all_pts; pts_out(j,1), pts_out(j,2), 1, pts_out(j,1)]; end
end
if ~isempty(pts_in)
    for j = 1:size(pts_in,1), all_pts = [all_pts; pts_in(j,1), pts_in(j,2), 2, pts_in(j,1)]; end
end
if size(all_pts,1) < 2, return; end
[~, idx] = sort(all_pts(:,4)); all_pts = all_pts(idx, :);
for j = 1:size(all_pts,1)-1
    p1 = all_pts(j,1:2); p2 = all_pts(j+1,1:2);
    t1 = all_pts(j,3); t2 = all_pts(j+1,3);
    if t1 == 2 && t2 == 2, continue; end
    mx = (p1(1)+p2(1))/2; my = (p1(2)+p2(2))/2;
    if pointInEllipse(mx, my, a_in, b_in), continue; end
    z1 = hpZ(p1(1),p1(2)); z2 = hpZ(p2(1),p2(2));
    pillars = [pillars; p1(1),p1(2),0,z1; p2(1),p2(2),0,z2];
    roof_segments{end+1} = [p1(1),p1(2),z1; p2(1),p2(2),z2];
end
end

function pts = lineEllipseIntersect(m, d, a_ell, b_ell)
A = 1/a_ell^2 + m^2/b_ell^2; B = 2*m*d/b_ell^2; Cc = d^2/b_ell^2 - 1;
disc = B^2 - 4*A*Cc;
if disc < 0, pts = [];
elseif disc < 1e-12, x = -B/(2*A); pts = [x, m*x+d];
else, x1 = (-B+sqrt(disc))/(2*A); x2 = (-B-sqrt(disc))/(2*A); pts = [x1, m*x1+d; x2, m*x2+d];
end
end

function inside = pointInEllipse(x, y, a_ell, b_ell)
inside = (x^2/a_ell^2 + y^2/b_ell^2) < 1;
end
